### PR TITLE
Inject a SPIR-V env when going from TritonIntelGPU to LLVM

### PR DESCRIPTION
**Pull request by FMarno**
The purpose of this is to allow the use of `triton_gen.sub_group_reduce`, since the verification for was added in #1223.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h [hunk 1]
```diff
@@ -169,6 +208,9 @@ class TritonGPUToLLVMPipelineManager {
     using namespace mlir;
     using namespace mlir::triton;
 
+    // should run before other patterns need the SPIRV-ENV attr
+    patterns.add<AddSPIRVEnvPattern>(&typeConverter.getContext(), benefit);
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h:212]: Can we define a higher benefit as a constant in the header where all benefits are defined and use it here?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 2]
```diff
@@ -65,7 +68,8 @@ struct ConvertTritonGPUToLLVM
   using ConvertTritonIntelGPUToLLVMBase::ConvertTritonIntelGPUToLLVMBase;
 
   void getDependentDialects(DialectRegistry &registry) const override {
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:70]: I'm surprised we have to do this. We may be defining the pass in a weird way. Not to be handled in this PR, ofc.

**Comment by victor-eds** [comment]
Adding the SPIR-V env LGTM. Issue I find is: I'm not sure we can avoid the information duplication (we have SPIR-V env and `triton_gpu` now) as LLVM dialect -> LLVM IR custom attribute translation requires attribute to have a `triton_gen` prefix name...

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h (hunk 1 above)
Comment (FMarno) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h]: Nice idea https://github.com/intel/intel-xpu-backend-for-triton/pull/1270/commits/4f8deae68537580b31ab2721f30933b70274de6d

**Review by victor-eds — APPROVED** [review]

File: include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h [hunk 3]
```diff
@@ -26,6 +26,7 @@ constexpr int patternBenefitDefault = 1;
 constexpr int patternBenefitPrioritizeOverLLVMConversions = 10;
 constexpr int patternBenefitClampOptimizedPattern = 20;
 constexpr int patternBenefitConvertLayoutOptimizedPattern = 20;
+constexpr int patternBenefitAddSPIRVEnv = 30;
```
Comment (whitneywhtsang) [inline include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h:29]: I see the benefit of having all the benefits at the same place, but having changes in common file, i.e., not under third_party/intel, could impact upstreaming. We are working in the direction if minimizing changes in common files, so at some point, all changes requires are in `third_party/intel`, then we could be a plugin to the upstream project.

Comment (victor-eds) [reply include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h:29]: Can we get this merged and find a place for this kind of stuff later or do you prefer to settle this now?

**Review by whitneywhtsang — APPROVED** [review]

File: include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h (hunk 3 above)
Comment (whitneywhtsang) [reply include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h:29]: We can get this PR merged first. Please open an issue to track that.

Comment (FMarno) [reply include/triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h:29]: Created #1273 for this

**Comment by whitneywhtsang** [comment]
> Adding the SPIR-V env LGTM. Issue I find is: I'm not sure we can avoid the information duplication (we have SPIR-V env and `triton_gpu` now) as LLVM dialect -> LLVM IR custom attribute translation requires attribute to have a `triton_gen` prefix name...

Can we enhance gpu-to-llvm-spv pass to translate those attributes?
@FMarno Please create a new issue to track the removal of the duplication.

**Comment by victor-eds** [comment]
> Can we enhance gpu-to-llvm-spv pass to translate those attributes?

That pass shouldn't be handling SPIR-V attributes. We should have GPu attributes with the same semantics that are lowered to other dialects. That's a pending task.

File: test/Conversion/intel/tritonintelgpu_to_llvm.mlir [hunk 4]
```diff
@@ -0,0 +1,5 @@
+// RUN: triton-opt %s --convert-triton-intel-gpu-to-llvm | FileCheck %s
+
+// COM: check that the spirv target env is inserted
+// CHECK: spirv.target_env{{.*}}#spirv.resource_limits<subgroup_size = 16>
```
Comment (etiotto) [inline test/Conversion/intel/tritonintelgpu_to_llvm.mlir:4]: Where is the spirv.target_env added? It should be to the module?

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h [hunk 5]
```diff
@@ -135,6 +136,44 @@ struct FuncOpConversion : public ConvertOpToLLVMPattern<triton::FuncOp> {
...
+  using mlir::OpRewritePattern<ModuleOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ModuleOp op,
+                                PatternRewriter &rewriter) const override {
+    if (spirv::lookupTargetEnv(op)) {
+      return failure();
+    }
+
+    int subgroupSize =
+        ::mlir::triton::gpu::TritonGPUDialect::getThreadsPerWarp(op);
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h:150]: [nit]: the :: in `::mlir` is not required.

File: test/Conversion/intel/tritonintelgpu_to_llvm.mlir (hunk 4 above)
Comment (victor-eds) [reply test/Conversion/intel/tritonintelgpu_to_llvm.mlir:4]: It is, but check's a bit lax. True.

Comment (etiotto) [reply test/Conversion/intel/tritonintelgpu_to_llvm.mlir:4]: ok then let's fix the CHECK to make it clear.

Comment (FMarno) [reply test/Conversion/intel/tritonintelgpu_to_llvm.mlir:4]: I'll open a quick PR for that

Comment (whitneywhtsang) [reply test/Conversion/intel/tritonintelgpu_to_llvm.mlir:4]: addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1286.

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h (hunk 5 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h:150]: addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1286.

**Pull request merged**